### PR TITLE
Reduce per-product work in listing swatch renderer (#39073)

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)

Two contained reductions of per-configurable work in the category listing swatch renderer. They do not remove the nine per-parent SQL statements the issue measures (each is intrinsically parent-scoped, and everything parent-independent is already memoized in EAV `Source\Table`, `Swatches\Helper\Data` and `SwatchAttributeCodes`), so this is a partial improvement on the axis the reporter is on, variations per configurable:

- `Renderer\Configurable::getConfigurableOptionsIds()` called `helper->getAllowAttributes($product)` inside the per-child loop, resolving the parent's configurable attributes once per variation. With 250 variations and 12 listed products that is 3000 resolutions per page. It is now resolved once per parent; the option id set is unchanged.
- `Helper\Data::loadVariationByFallback()` loaded the whole matching child collection and used `getFirstItem()`. It now sets `setPageSize(1)`. This path runs per parent from `Listing\Configurable::_getAdditionalConfig()` whenever a swatch attribute is in the layered navigation query.

Also fixes the pre-existing Static Tests warnings in the touched block (`public const`, `@see` on deprecated members, explicit `return null`).

### Fixed Issues (if relevant)

1. Partially addresses magento/magento2#39073

### Manual testing scenarios (*)

1. Category page with configurable products and swatches enabled: swatches, prices and images render as before.
2. Filter that category by a swatch attribute (layered navigation): the preselected variation image is unchanged.
3. Profile the page: `Swatches\Helper\Data::getAllowAttributes` call count drops from children x parents to parents.

### Contribution checklist (*)

- [x] Pull request has a meaningful description of its purpose
- [x] All commits are accompanied by meaningful commit messages
- [x] All new or changed code is covered with unit/integration tests (if applicable)
- [x] README.md files for modified modules are updated and included in the pull request if any README.md predefined sections require an update
- [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php [hunk 1]
```diff
@@ -420,17 +423,21 @@ protected function isProductHasImage(Product $product, $imageType)
...
-                $productAttributeId = $productAttribute->getId();
-                if (isset($attributeData[$productAttributeId])) {
-                    $attributeValue = $product->getData($productAttribute->getAttributeCode());
-                    if ($attributeValue !== null) {
-                        $ids[$attributeValue] = 1;
-                    }
+            foreach ($attributeCodes as $attributeCode) {
+                $attributeValue = $product->getData($attributeCode);
+                if ($attributeValue !== null) {
+                    $ids[$attributeValue] = 1;
```
Comment (ilnytskyi) [inline app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php:440]: If only keys needed from `$ids` - the subsequent `array_keys` fcall can be also eliminated after changing `$ids[$attributeValue] = 1;` to `$ids[] = $attributeValue;`.

Comment (lbajsarowicz) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php:440]: Checked this against the data flow before changing anything, and I am going to keep `array_keys($ids)`.

`$attributeValue` does repeat across the loop. It is nested per allowed product and per swatch attribute code, and sibling variations routinely share a value — ten children in "Red", five in size "L". `$ids[$attributeValue] = 1` is a deliberate dedup, so `array_keys($ids)` returns one entry per distinct option, not one per product.

That dedup is essential downstream. The result goes to `SwatchHelper::getSwatchesByOptionsId()`, which decides whether the cache already holds everything with:

```php
$swatches = $this->getCachedSwatches($optionIds);
if (count($swatches) !== count($optionIds)) {
```

and `getCachedSwatches()` is `array_intersect_key($this->swatchesCache, array_combine($optionIds, $optionIds))`. Both `array_combine` and `array_intersect_key` key by option id, so `count($swatches)` can never exceed the number of *distinct* ids. Feed duplicates in and `count($optionIds)` is permanently larger, the check never passes even with a fully warm cache, and every call falls through to `addFilterByOptionsIds()` and a fresh collection query.

That would defeat swatch caching on any listing with repeated swatch values, which is the normal case, and it works against exactly what this PR is for. No test catches it today because the unit fixtures use small already-distinct value sets.

Separately, on the red Semantic Version Checker: it reports one MAJOR, `M121 [protected] Method return typing changed` on `getSwatchProductImage`. The method is `protected`, has no declared return type, and could already return `null` — the pre-patch code simply fell off the end when neither branch matched while the docblock claimed `@return string`. This PR adds the explicit `return null;` and corrects the docblock to `@return string|null`. Runtime behaviour is unchanged; SVC is flagging a documentation correction. I would rather leave the docblock accurate than restore a wrong one to silence the check, but say the word if you want it reverted.

Comment (ilnytskyi) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php:440]: You’re right deduplication needs to be preserved, so `$ids[] = $attributeValue` isn’t equivalent.
It can instead use `$ids[$attributeValue] = $attributeValue` and return $ids directly.
The downstream code consumes the values and does not require sequential keys.
The gain is very small, though, so either implementation is fine.

File: app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php [hunk 2]
```diff
@@ -397,6 +398,8 @@ protected function getSwatchProductImage(Product $childProduct, $imageType)
         if (!empty($swatchImageId) && !empty($imageAttributes['type'])) {
             return $this->imageUrlBuilder->getUrl($childProduct->getData($imageAttributes['type']), $swatchImageId);
         }
+
+        return null;
```
Comment (Den4ik) [inline app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php:402]: @lbajsarowicz, Do you think you could avoid changing the return type of the protected method so that the Semantic Version Check passes without errors? This would allow to include these changes in a minor release without requiring approval from the Architect.

Comment (lbajsarowicz) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php]: Do you suggest empty string?

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php]: @ilnytskyi thanks, agreed that `$ids[$attributeValue] = $attributeValue` would work too. I'll keep `array_keys()` here to keep the diff small, since the gain is marginal.

File: app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php]: @Den4ik done: the docblock is back to `@return string`, so the method's declared return typing matches 2.4-develop. The explicit `return null;` fallback stays because PHPStan fails the changed file without it (`should return string but return statement is missing`); runtime behaviour is the same as before, where the method already fell through to an implicit null.

Comment (Den4ik) [reply app/code/Magento/Swatches/Block/Product/Renderer/Configurable.php]: Thanks @lbajsarowicz
